twiliohandle interruption

aashsach opened the pull request:
added  bot interruption in twilio serilaizer.
it will clear the web socket buffer whenever there is an interruption.

aconchillo commented:
Just for reference. There is another approach here: https://github.com/pipecat-ai/pipecat/pull/395

aconchillo commented on src/pipecat/transports/network/fastapi_websocket.py:
```diff
@@ -114,37 +122,66 @@ async def write_raw_audio_frames(self, frames: bytes):
+    async def write_websocket_clear_event(self, frame: StartInterruptionFrame):
+        payload = self._params.serializer.serialize(frame)
+        if payload and self._websocket.client_state == WebSocketState.CONNECTED:
+            await self._websocket.send_text(payload)
```
It seems this is not used and probably not necessary. This could be achieved from any processor by sending a `StartInterruptionFrame`. In general, if we want to achieve something in Pipecat we do it through frames, instead of calling functions directly.

aconchillo replied:
You are totally right. I forgot this class was based on `BaseOutputTransport` (it originally wasn't at some point).

aconchillo commented on src/pipecat/transports/network/fastapi_websocket.py:
```diff
@@ -114,37 +122,66 @@ async def write_raw_audio_frames(self, frames: bytes):
+            self._create_push_task()
+            # Let's send a bot stopped speaking if we have to.
+            if self._bot_speaking:
+                await self._bot_stopped_speaking()
```
It seems `_handle_interruptions()` is not used.

aconchillo replied:
Yes, sorry, I completely forgot this were subclasses of the base transports.

aconchillo commented on src/pipecat/serializers/twilio.py:
```diff
@@ -28,22 +28,24 @@ def __init__(self, stream_sid: str, params: InputParams = InputParams()):
+
+        if isinstance(frame, StartInterruptionFrame):
+            answer = {"event": "clear", "streamSid": self._stream_sid}
+            return json.dumps(answer)
```
I think I like this approach. In this case probably better to use `UserStartedSpeaking` instead. The `StartInterruptionFrame` is more an internal/system frame (we use it to cancel tasks and more system specific things). `UserStartedSpeaking` will be pushed just after the interruption, so it will have the same effect.

aashsach replied:
using StartInterruptionFrame, since it is natively handled by process_from of base transport output processor, and it is usually handled by all processors separately, following similar pattern

aconchillo commented:
This is great, thank you! I added a few comments. Also, I see a bunch of formatting changes, not sure why. Would it be possible to avoid those changes?

aashsach commented on src/pipecat/transports/network/fastapi_websocket.py:
```diff
@@ -97,7 +97,8 @@ async def write_raw_audio_frames(self, frames: bytes):
         while len(self._websocket_audio_buffer) >= self._params.audio_frame_size:
             frame = AudioRawFrame(
-                audio=self._websocket_audio_buffer[:self._params.audio_frame_size],
+                audio=self._websocket_audio_buffer[:
```
this is auto formatted by autopep8 extension in vscode

aashsach commented:
> This is great, thank you! I added a few comments. Also, I see a bunch of formatting changes, not sure why. Would it be possible to avoid those changes?

thanks for the pr.
reverted formatting issues, added comments

aashsach commented:
@aconchillo would you be able to look at this PR again

aconchillo commented:
> @aconchillo would you be able to look at this PR again

Yes, very soon! And thank you for all your work! :heart:

aconchillo commented:
Sorry for the delay. I took a deeper look this time. This is what I would suggest:

- Don't override `_handle_interruptions`. This can internally change in the future and we would need to update this class as well.
- Override `process_frame()` instead.
- Use `UserStartedSpeakingFrame` for serialization.
- Rename `write_websocket_clear_event` to `_write_frame`. This way we can add more frames to the `process_frame` and serialize them and write them to the websocket.

So, you would end up with something as simple as:

```
async def process_frame(self, frame: Frame, direction: FrameDirection):
     await super().process_frame(frame, direction)
     
     if isinstance(frame, UserStartedSpeakingFrame):
         await self._write_frame(frame)
         
async def _write_frame(self, frame: Frame):
    payload = self._params.serializer.serialize(frame)
    if payload and self._websocket.client_state == WebSocketState.CONNECTED:
        await self._websocket.send_text(payload)
```

aashsach commented:
> Sorry for the delay. I took a deeper look this time. This is what I would suggest:
> 
> * Don't override `_handle_interruptions`. This can internally change in the future and we would need to update this class as well.
> * Override `process_frame()` instead.
> * Use `UserStartedSpeakingFrame` for serialization.
> * Rename `write_websocket_clear_event` to `_write_frame`. This way we can add more frames to the `process_frame` and serialize them and write them to the websocket.
> 
> So, you would end up with something as simple as:
> 
> ```
> async def process_frame(self, frame: Frame, direction: FrameDirection):
>      await super().process_frame(frame, direction)
>      
>      if isinstance(frame, UserStartedSpeakingFrame):
>          await self._write_frame(frame)
>          
> async def _write_frame(self, frame: Frame):
>     payload = self._params.serializer.serialize(frame)
>     if payload and self._websocket.client_state == WebSocketState.CONNECTED:
>         await self._websocket.send_text(payload)
> ```

thanks!
so to override proces_frame, we would need to subclass fast API WebSocket transport, the recommendation is to create a new Twilio transport that inherits from fast API WebSockets and overrides relevant things.

aconchillo commented:
> > Sorry for the delay. I took a deeper look this time. This is what I would suggest:
> > 
> > * Don't override `_handle_interruptions`. This can internally change in the future and we would need to update this class as well.
> > * Override `process_frame()` instead.
> > * Use `UserStartedSpeakingFrame` for serialization.
> > * Rename `write_websocket_clear_event` to `_write_frame`. This way we can add more frames to the `process_frame` and serialize them and write them to the websocket.
> > 
> > So, you would end up with something as simple as:
> > ```
> > async def process_frame(self, frame: Frame, direction: FrameDirection):
> >      await super().process_frame(frame, direction)
> >      
> >      if isinstance(frame, UserStartedSpeakingFrame):
> >          await self._write_frame(frame)
> >          
> > async def _write_frame(self, frame: Frame):
> >     payload = self._params.serializer.serialize(frame)
> >     if payload and self._websocket.client_state == WebSocketState.CONNECTED:
> >         await self._websocket.send_text(payload)
> > ```
> 
> thanks! so to override proces_frame, we would need to subclass fast API WebSocket transport, the recommendation is to create a new Twilio transport that inherits from fast API WebSockets and overrides relevant things.

I believe you should be able to override directly in this class. If the given serializer is the twilio one the check `if payload` will take care of everything. If it's another serializer, the serializer might return None in which case nothing will happen.

aashsach commented:
@aconchillo I have made the recommended changes.
i have one question though, daily output does not handle UserStartedSpeakingFrame, it resets states via interruption frame, why should Twilio transport be different?

aconchillo commented:
> @aconchillo I have made the recommended changes. i have one question though, daily output does not handle UserStartedSpeakingFrame, it resets states via interruption frame, why should Twilio transport be different? in addition, it won't spend botstopspeaking frame without actually handling interruptions. wont this break at multiple places

`StartInterruptionFrame` is a system frame (base class of `SystemFrame`. System frames are not queued by any processor and are sent directly downstream. The downstream processors use `StartInterruptionFrame` to usually cancel tasks or any other low level thing they do.

After `StartInterruptionFrame`, `UserStartedSpeakingFrame` is pushed downstream. That's a `ControlFrame` and control frames are ordered together with `DataFrame`. These frames can be serialized, like you did, and sent through the transport if necessary.

It doesn't really matter if `UserStartedSpeakingFrame` is not handled by the base transport, in the base transport we just have don't use for it. However, in the subclass we can use it to send the websocket message we need to Twilio. So, by override `process_frame()` we have the option to do more things on the incoming frame.

If the FastAPI websocket subclass had some task that needed to be cancel, we would use `StartInterruptionFrame`, but we would still use `UserStartedSpeakingFrame` to send the websocket message.

Does that make more sense?

aconchillo commented:
> in addition, it won't spend botstopspeaking frame without actually handling interruptions. wont this break at multiple places

Not sure what you mean here. The `BotStoppedSpeaking` frame is pushed by the base output upstream and downstream. I don't think this implementation would affect that.

aconchillo commented:
Changes look great to me. I think this should work! I assume it keeps working? :sweat_smile:

aashsach commented:
> > @aconchillo I have made the recommended changes. i have one question though, daily output does not handle UserStartedSpeakingFrame, it resets states via interruption frame, why should Twilio transport be different? in addition, it won't spend botstopspeaking frame without actually handling interruptions. wont this break at multiple places
> 
> `StartInterruptionFrame` is a system frame (base class of `SystemFrame`. System frames are not queued by any processor and are sent directly downstream. The downstream processors use `StartInterruptionFrame` to usually cancel tasks or any other low level thing they do.
> 
> After `StartInterruptionFrame`, `UserStartedSpeakingFrame` is pushed downstream. That's a `ControlFrame` and control frames are ordered together with `DataFrame`. These frames can be serialized, like you did, and sent through the transport if necessary.
> 
> It doesn't really matter if `UserStartedSpeakingFrame` is not handled by the base transport, in the base transport we just have no use for it. However, in the websocket subclass we can use it to send the websocket message we need to Twilio. So, by overriding `process_frame()` we have the option to do more things on the incoming frame.
> 
> If the FastAPI websocket subclass had some task that needed to be cancel, we would use `StartInterruptionFrame`, but we would still use `UserStartedSpeakingFrame` to send the websocket message.
> 
> Does that make more sense?
> 
> 

Hi thanks for the detailed explanation.
That makes perfect sense.
Where  I am confused is this concept of handling interruption frame to cancel output task. Egz in daily this is used to stop sending audio frame to output and it works because daily is more synchronised with when the audio is sent and when it's is spoken.
In twilio, I want to do the same functionality, cancel whatever vot is supposed to speak whenever there is interruption. However this needs to happen at twilio 's end because when audio is sent vs when it is spoken,  is out of sync.
So essentially we are not serialising a data frame, we are cancelling the bot speaking task via sending a message, shouldn't that be triggered via system frame so that bot stops as quickly as possible instead of waiting for message to propagate via queues.

aconchillo commented:
> Hi thanks for the detailed explanation. That makes perfect sense. Where I am confused is this concept of handling interruption frame to cancel output task. Eg in daily this is used to stop sending audio frame to output and it works because daily is more synchronised with when the audio is sent and when it is spoken. In twilio, I want to do the same functionality, cancel whatever bot is supposed to speak whenever there is interruption. However this needs to happen at twilio 's end because when audio is sent vs when it is spoken, is out of sync. So essentially we are not serialising a data frame, we are cancelling the bot speaking task via sending a message, shouldn't that be triggered via system frame so that bot stops as quickly as possible instead of waiting for message to propagate via queues.

`UserStartedSpeeaking` is always sent just after `StartInterruptionFrame`, so it wouldn't really matter. But yes, thinking more about it, I believe you might be right. Let's switch to `StartInterruptionFrame`. We can always change it in the future, right now both would work.

aashsach commented:
@aconchillo incorporated changes!

aconchillo commented:
> @aconchillo incorporated changes!

Thank you! LGTM! Merged!
